### PR TITLE
http2: another round of updates

### PR DESCRIPTION
**Pull request by jasnell**
Starting another round of updates...

##### Checklist

- [ ] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [ ] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [ ] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/CONTRIBUTING.md#commit-message-guidelines)

### CODE REVIEW

File: lib/internal/http2/core.js [hunk 1]
```diff
@@ -1026,7 +1026,6 @@ function streamOnceReady() {
 // on the client side it is used to transmit request data.
 class Http2Stream extends Duplex {
   constructor(session, options) {
-    options.allowHalfOpen = true;
```
Comment (mcollina) [inline lib/internal/http2/core.js]: why?

File: src/node_http2.cc [hunk 2]
```diff
@@ -912,38 +769,33 @@ void Http2Session::OnHeaders(Nghttp2Stream* stream,
...
+      argv[j * 2 + 1] = value_str;
+      headers = item->next;
+      j++;
     }
-    headers = item->next;
-  }
+    if (j > 0) {
+      fn->Call(env()->context(), holder, j * 2, argv).ToLocalChecked();
+    }
+  } while (headers != nullptr);
```
Comment (mcollina) [inline src/node_http2.cc]: can you add some comments explaining how all of this works? It's a bit obscure and it might help in the future.

**Review by mcollina — APPROVED** [review]
LGTM with nits

File: lib/internal/http2/core.js (hunk 1 above)
Comment (jasnell) [reply lib/internal/http2/core.js]: I may end up pulling this particular change out... it appeared to have been causing an issue with some clients but further testing showed some other issues.

**Comment by jasnell** [comment]
Landed!

**Pull request closed without merging**